pycloudlib/instance.py

```diff
@@ -370,29 +371,28 @@ def _tmpfile(self):
         self._tmp_count += 1
         return path
 
-    def _wait_for_execute(self):
+    def _wait_for_boot_id(self):
         """Wait until we can execute a command in the instance."""
-        self._log.debug('_wait_for_execute to complete')
+        self._log.debug('_wait_for_boot_id to complete')
 
-        test_instance_command = "whoami"
-        result = self.execute(test_instance_command)
-        if result.failed:
-            retries = 10
-            while retries:
-                result = self.execute(test_instance_command)
-
-                if result.ok:
+        boot_id_command = "cat /proc/sys/kernel/random/boot_id"
```

Comment:
Thinking about this further, part of the reason `whoami` is used here is that it's present on pretty much any Unix system: failures to run it almost certainly mean that we don't have access to the system. This replacement will never succeed on non-Linux systems, such as *BSD, where we'll block for 5 minutes and then fail.

We also won't have `self.boot_id` set on instances which aren't `wait`'d for at all, and it will be stale on subsequent boots of instances which are rebooted by means other than pycloudlib: `cc_power_state` or even a simple `.execute("sudo reboot")`. This will make using it tricky: it won't necessarily be present when we `restart`, so we'd have to determine it at that point regardless. I wonder if this indicates that instead of storing it at the instance-level we should contain it to the `restart` path, something like: determine current boot ID, trigger a restart, then call `self.wait` in a loop which we break out of once the post-`wait` boot ID doesn't match?

(As an aside, we should consider if whatever we do here allows us to drop (or substantially simplify) the reboot detection code in `test_power_state_change` in cloud-init; if we can't readily use it there, then that _may_ mean we have not found the right abstraction.)

Reply:
I thought it might be possibly useful in other ways of checking boot too, but you're right in that it'd be a lot more work to keep the boot id current. So I agree, let's restrict it to restart.

Your parenthetical is right too. The boot detecting code was initially added there because of this, so a good test of this should be to remove that code.